hashes/src/internal_macros.rs

```diff
@@ -235,3 +235,19 @@ macro_rules! hash_type {
     };
 }
 pub(crate) use hash_type;
+
+/// Requires `HashEngine` trait to be in scope.
+macro_rules! write_for_hash_engine_impl {
+    () => {
+        #[cfg(feature = "bitcoin-io")]
+        bitcoin_io::impl_write!(
+            HashEngine,
+            |us: &mut HashEngine, buf| {
+                us.input(buf);
+                Ok(buf.len())
+            },
+            |_us| { Ok(()) }
+        );
+    }
+}
```

Comment:
I meant to just dump this into `hash_type` macro (not to be confused with `hash_newtype`).

Reply:
I don't think it belongs in there. Since this is all internal stuff anyways its not a super big issue, right?

Reply:
We should rename the macro to `hash_mod` :) All modules should be consistent and that should be the main purpose of the macro.

Reply:
Agreed, leaving for another day though. This PR simply reduces code duplication without further tackling the `hashes` macros, I think there is still a bunch of things to improve in the macros.

Reply:
We can also just close this if you think its not a useful step forward.

Reply:
The main reason for the macro is consistency, not deduplication. I think we should just rename the macro as a first step and then make sure to move stuff into it.